src/main/java/org/wildfly/security/auth/spi/Attributes.java

```diff
@@ -576,13 +592,66 @@ default boolean isEmpty() {
     }
 
     /**
-     * Get the collection of values for the given key.  The result may implement {@link SetEntry} if the values
-     * are distinct (for example, a role or group set).
+     * Returns a read-only instance of this instance.
      *
-     * @param key the attribute name
-     * @return the (possibly empty) attribute value collection
+     * @return a read-only instance of this instance.
      */
-    Entry get(String key);
+    default Attributes asReadOnly() {
+        return new Attributes() {
+            @Override
+            public Collection<Entry> entries() {
+                Collection<Entry> entries = Attributes.this.entries();
+                Iterator<Entry> iterator = entries.iterator();
```

Comment:
This has to be at line 608 above ``return new Iterator...`` else you'll only be able to iterate one time.